File: Python/63_Unique_Paths.py
class Solution(object):
    def uniquePathsWithObstacles(self, obstacleGrid):
        """
        :type obstacleGrid: List[List[int]]
        :rtype: int
        """

        def isObs(obstacleGrid, row, col):
            if obstacleGrid[row][col] == 1:
                return True
            return False

        m = len(obstacleGrid)
        n = len(obstacleGrid[0])

        memo = [[0 for _ in range(m)] for _ in range(n)]
        memo[m-1][n-1] = 1

        for row in range(m-1,-1,-1):
            for col in range(n-1,-1,-1):
                if isObs(obstacleGrid,row,col) == False:
                    if row < m-1:
                        memo[row][col] += memo[row+1][col]
                    if col < n-1:
                        memo[row][col] += memo[row][col+1]

        return memo[0][0]            

# algorithm: memoization
# unique paths can be calculated from the bottom-right
# must check if the cell is not an obstable
        

File: C++/63_UniquePaths.cpp
using namespace std;
#include <vector>

class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
        int rMax = obstacleGrid.size();
        int cMax = obstacleGrid[0].size();
        long long grid[rMax][cMax];

        for (int r=0;r<rMax;r++){
            for (int c=0;c<cMax;c++) grid[r][c] = 0;
        }

        if (obstacleGrid[0][0] == 1 || obstacleGrid[rMax-1][cMax-1] == 1) return 0;

        grid[rMax-1][cMax-1] = 1;

        for(int r=rMax-1;r>-1;r--){
            for(int c=cMax-1;c>-1;c--){
                if (!obstacleGrid[r][c]) {
                    if (r+1 < rMax && !obstacleGrid[r+1][c]) grid[r][c] += grid[r+1][c];
                    if (c+1 < cMax && !obstacleGrid[r][c+1]) grid[r][c] += grid[r][c+1];
                }
            }
        }
        
        return grid[0][0];
    }
};